response.go

```diff
@@ -0,0 +1,248 @@
+	// certificate (including the signature)
+	CertHash []byte
+
+	// IssuerSerial holds the issuer and serialNumber of the certificate.
+	// When it is not present, the SignerIdentifier field in the SignerInfo
+	// will be used.
+	IssuerSerial issuerAndSerial `asn1:"optional"`
+}
+
+// issuerAndSerial holds the issuer name and serialNumber of the certificate
+//
+// Reference: RFC 5035 4
+//
+//	IssuerSerial ::= SEQUENCE {
+//		issuer                   GeneralNames,
+//		serialNumber             CertificateSerialNumber }
+type issuerAndSerial struct {
+	IssuerName   generalNames
+	SerialNumber *big.Int
+}
+
+// generalNames holds the issuer name of the certificate.
+//
+// Reference: RFC 3280 4.2.1.7
+//
+// GeneralNames ::= SEQUENCE SIZE (1..MAX) OF GeneralName
+//
+//	GeneralName ::= CHOICE {
+//		otherName                       [0]     OtherName,
+//		rfc822Name                      [1]     IA5String,
+//		dNSName                         [2]     IA5String,
+//		x400Address                     [3]     ORAddress,
+//		directoryName                   [4]     Name,
+//		ediPartyName                    [5]     EDIPartyName,
+//		uniformResourceIdentifier       [6]     IA5String,
+//		iPAddress                       [7]     OCTET STRING,
+//		registeredID                    [8]     OBJECT IDENTIFIER }
+type generalNames struct {
+	Name asn1.RawValue `asn1:"optional,tag:4"`
+}
+
+// MarshalBinary encodes the response to binary form.
+// This method implements encoding.BinaryMarshaler.
+//
+// Reference: https://pkg.go.dev/encoding#BinaryMarshaler
+func (r *Response) MarshalBinary() ([]byte, error) {
+	if r == nil {
+		return nil, errors.New("nil response")
+	}
+	return asn1.Marshal(*r)
+}
+
+// UnmarshalBinary decodes the response from binary form.
+// This method implements encoding.BinaryUnmarshaler.
+//
+// Reference: https://pkg.go.dev/encoding#BinaryUnmarshaler
+func (r *Response) UnmarshalBinary(data []byte) error {
+	_, err := asn1.Unmarshal(data, r)
+	return err
+}
+
+// ValidateStatus validates the response.Status
+//
+// Reference: RFC 3161 2.4.2
+func (r *Response) ValidateStatus() error {
+	if r.Status.Status != pki.StatusGranted && r.Status.Status != pki.StatusGrantedWithMods {
+		failureInfo, err := r.Status.ParseFailInfo()
+		if err != nil {
+			return &InvalidResponseError{Msg: fmt.Sprintf("invalid response with status code %d: %s", r.Status.Status, r.Status.Status.String())}
+		}
+		return &InvalidResponseError{Msg: fmt.Sprintf("invalid response with status code %d: %s. Failure info: %s", r.Status.Status, r.Status.Status.String(), failureInfo)}
+	}
+	return nil
+}
+
+// SignedToken returns the timestamp token with signatures.
+// Callers should invoke Verify to verify the content before comsumption.
+func (r *Response) SignedToken() (*SignedToken, error) {
+	if err := r.ValidateStatus(); err != nil {
+		return nil, err
+	}
+	return ParseSignedToken(context.Background(), r.TimeStampToken.FullBytes)
+}
```

Comment:
validate that if TimeStampToken is not present, the PKIFailureInfo is present

Reply:
Do you mean `PKIStatusInfo`? because `PKIFailureInfo` is `OPTIONAL` defined by [RFC 3161 2.4.2](https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.2).

According to RFC 3161, we do have the check for `PKIStatusInfo`, if its `status` is `zero or one`, then we require `a TimeStampToken MUST be present`. Otherwise, `SignedToken()` would be failed. If `status` is NOT `zero or one`, we break early and check if `PKIFailureInfo` is present. You can find the above logic at `Response.ValidateStatus()`.